cloudflare/resource_cloudflare_api_token.go

```diff
@@ -22,9 +22,10 @@ func resourceCloudflareApiToken() *schema.Resource {
 				Elem:     &schema.Schema{Type: schema.TypeString},
 			},
 			"permission_groups": {
-				Type:     schema.TypeList,
```

Comment:
as we are going from a list => set, we will need to 
1. bump the schema version 
2. provide a state migration (see https://www.terraform.io/docs/extend/resources/state-migration.html) as we will be changing the internal representation of the data

Reply:
Is this really necessary?

I tested this:
1. create resource with current provider version
2. upgrade to fixed provider
3. run `terraform apply`

No changes detected.

Reply:
Because state presentation of a list or a set is the same - json array.

Reply:
`TypeList` and `TypeSet` definitely aren't the same state representation as [sets are stored indexed by a hash calculated based on the attributes in the set](https://www.terraform.io/docs/extend/schemas/schema-types.html#typeset) whereas [list items are stored as a zero indexed items](https://www.terraform.io/docs/extend/schemas/schema-types.html#typelist). I'm unsure why you're not seeing a difference here but we may need to dig into why as I've definitely been bit by this in the past.

Reply:
Sorry for late response. But I said that because I checked the state before and after the fix. Even at links you posted are examples, both of them with slices. I think terraform is doing hash keys on the fly and not storing them in the state file.

Just do the test please.

For the given resource
```hcl
resource "cloudflare_api_token" "test" {
  name = "test"

  policy {
    effect = "allow"
    permission_groups = [
      "3030687196b94b638145a3953da2b699",
      "da6d2d6f2ec8442eaadda60d13f42bca",
      "e6d2666161e84845a636613608cee8d5",
      "ed07f6c337da4195b4e72a1fb2c6bcae",
      "1af1fa2adc104452b74a9a3364202f20",
      "4755a26eedb94da69e1066d98aa820be",
    ]
    resources = { "com.cloudflare.api.account.*" = "*" }
  }
}
```


Current provider state
```json
{
  "version": 4,
  "terraform_version": "0.13.7",
  "serial": 1,
  "lineage": "f3059c1d-de72-cc99-752d-da6446426758",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "cloudflare_api_token",
      "name": "test",
      "provider": "provider[\"registry.terraform.io/terraform-providers/cloudflare\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "condition": [],
            "id": "e8712ce735c7353116fe7586dcf5bacb",
            "issued_on": "2021-07-19T12:26:40Z",
            "modified_on": "2021-07-19T14:26:42.035223871+02:00",
            "name": "uros-test",
            "policy": [
              {
                "effect": "allow",
                "permission_groups": [
                  "3030687196b94b638145a3953da2b699",
                  "da6d2d6f2ec8442eaadda60d13f42bca",
                  "e6d2666161e84845a636613608cee8d5",
                  "ed07f6c337da4195b4e72a1fb2c6bcae",
                  "1af1fa2adc104452b74a9a3364202f20",
                  "4755a26eedb94da69e1066d98aa820be"
                ],
                "resources": {
                  "com.cloudflare.api.account.*": "*"
                }
              }
            ],
            "status": "active",
            "value": "REDACTED"
          },
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
```


Fixed provider state
```json
{
  "version": 4,
  "terraform_version": "0.13.7",
  "serial": 1,
  "lineage": "fb23297b-3d7b-2ce0-21fc-f1e9a7a75219",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "cloudflare_api_token",
      "name": "test",
      "provider": "provider[\"registry.terraform.io/terraform-providers/cloudflare\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "condition": [],
            "id": "8a712ce0983ed8b47adc2710e0094ef6",
            "issued_on": "2021-07-19T12:28:08Z",
            "modified_on": "2021-07-19T14:28:10.23180693+02:00",
            "name": "uros-test",
            "policy": [
              {
                "effect": "allow",
                "permission_groups": [
                  "1af1fa2adc104452b74a9a3364202f20",
                  "3030687196b94b638145a3953da2b699",
                  "4755a26eedb94da69e1066d98aa820be",
                  "da6d2d6f2ec8442eaadda60d13f42bca",
                  "e6d2666161e84845a636613608cee8d5",
                  "ed07f6c337da4195b4e72a1fb2c6bcae"
                ],
                "resources": {
                  "com.cloudflare.api.account.*": "*"
                }
              }
            ],
            "status": "active",
            "value": "REDACTED"
          },
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
```

You can see that state representation of both is json array. And as said already said, I tested creating resource and then upgrading provider. It works like a charm.

I will do it proper way if you convince me that current approach is not working, but sorry, I will no do something that is unnecessary IMO.